packages/core/src/browser/frontend-application-state.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable } from 'inversify';
+import { Emitter, Event } from '../common/event';
+import { Deferred } from '../common/promise-util';
+
+export type FrontendApplicationState =
+    'init'
+    | 'started_contributions'
+    | 'attached_shell'
+    | 'initialized_layout'
+    | 'ready'
+    | 'closing_window';
+
+@injectable()
+export class FrontendApplicationStateHandler {
```

Comment:
I think `Handler`is not the right term, here. Something that registers for statechange events would be called a handler. Something like 'service' or 'manager' would be more in line with how we name these kind of things.

Reply:
Or Provider